### ISSUE TITLE
Ubuntu 20.04 [amd64] kaltura-nginx packaging 'conflict' issue

### ISSUE THREAD

**Issue by gnanet** [kaltura/nginx-vod-module#1407]
**Please note, i was able to apply my use-case to the installation, but for other users the notes below may be of help.**

Followed the steps for the installation as described in [Debian/Ubuntu deb package](https://github.com/kaltura/nginx-vod-module#debianubuntu-deb-package) and have some issues to note:

First a side note: pre-dependency `kaltura-ffmpeg` installs an older version 4.2.2 than the system 4.2.7, and it does not declare a conflict, but adds its library path to ldconfig, so the version difference causing a warn in the system's 4.2.7 ffmpeg.

The package `kaltura-nginx` declares a conflict with `nginx` and that it would provide `httpd, nginx`, so the system removes the installed `nginx`, including simple management possibilities of  `nginx -t && nginx -s reload` to check and reload its config, because the kaltura-own nginx binary is not on PATH.
Also the postinst dialog does not highlight the change about the sugested port 88 from the standard port 80, which is not an obvious change, and a small vision deficiency makes hard to find why the webservice is not reachable on its usual port, which one would expect, if the package declares to provide httpd/nginx.

To sum up the problem:
A) If the kaltura-nginx package is intended to **only do transcode** on the fly, **it should not declare conflicts: and provides:** and can use port 88 and 8443 for example
B) If the kaltura-nginx package is meant to **replace current nginx, it should use ports 80 and 443**, and at least symlink the nginx binary to /usr/sbin/nginx

**Comment by jessp01** [comment]
Hi @gnanet ,

> First a side note: pre-dependency kaltura-ffmpeg installs an older version 4.2.2 than the system 4.2.7, and it does not declare a conflict, but adds its library path to ldconfig, so the version difference causing a warn in the system's 4.2.7 ffmpeg.

We intentionally do not declare `kaltura-ffmpeg` as conflicting with the distro's `ffmpeg` package because one may want both installed, each for different purposes (we need a specific version of FFmpeg for the VOD module, built with certain configuration flags). Since `/opt/kaltura/nginx/sbin/nginx` has a runtime dependency on FFMpeg's SOs:

```
# ldd /opt/kaltura/nginx/sbin/nginx 2>&1 |grep ffmpeg
	libavcodec.so.58 => /opt/kaltura/ffmpeg-4.2.2/lib/libavcodec.so.58 (0x00007fc13f131000)
	libavutil.so.56 => /opt/kaltura/ffmpeg-4.2.2/lib/libavutil.so.56 (0x00007fc13ee0f000)
	libswscale.so.5 => /opt/kaltura/ffmpeg-4.2.2/lib/libswscale.so.5 (0x00007fc13e789000)
	libavfilter.so.7 => /opt/kaltura/ffmpeg-4.2.2/lib/libavfilter.so.7 (0x00007fc13e0e3000)
	libswresample.so.3 => /opt/kaltura/ffmpeg-4.2.2/lib/libswresample.so.3 (0x00007fc13d2f4000)
	libpostproc.so.55 => /opt/kaltura/ffmpeg-4.2.2/lib/libpostproc.so.55 (0x00007fc139298000)
	libavformat.so.58 => /opt/kaltura/ffmpeg-4.2.2/lib/libavformat.so.58 (0x00007fc138dbc000)

```
We add it to LD conf:
```
# cat /etc/ld.so.conf.d/kaltura-ffmpeg.conf
/opt/kaltura/ffmpeg-4.2.2/lib
```

And indeed, if you have another `ffmpeg` package of a similar version (which means it will depend on SOs with the same names), you may get a warning. In such a case, we leave it to you to adjust the LD_LIBRARY_PATH according to your needs.

> The package kaltura-nginx declares a conflict with nginx and that it would provide httpd, nginx, so the system removes the installed nginx, including simple management possibilities of nginx -t && nginx -s reload to check and reload its config, because the kaltura-own nginx binary is not on PATH.

The `nginx` and `kaltura-nginx` packages can technically co-exist as they are installed onto different paths, on the other hand, since `kaltura-nginx` typically provides everything `nginx` from the official repo provides (and more) and seeing how Nginx, starting from version 1.9.11, supports dynamic modules, thus allowing you to build and load any additional SOs you may require, I do not feel there's a point for them to co-exist. Unlike with `ffmpeg` CLI, I cannot think of a use case that would require two separate Nginx versions to run in parallel. We do not add `/opt/kaltura/nginx/sbin` to PATH since it is not needed in order to control the Nginx daemon supplied with this package. You can use:
```
# /etc/init.d/kaltura-nginx $ACTION
```
or:
```
# service kaltura-nginx $ACTION
```
For example:
```
# service kaltura-nginx configtest
```
Of course, if one is so inclined, adding `/opt/kaltura/nginx/sbin` to `PATH` is quite easy. Also, using the full path will also work, to wit:
```
# /opt/kaltura/nginx/sbin/nginx -t
nginx: the configuration file /opt/kaltura/nginx/conf/nginx.conf syntax is ok
nginx: configuration file /opt/kaltura/nginx/conf/nginx.conf test is successful
```

> Also the postinst dialog does not highlight the change about the sugested port 88 from the standard port 80, which is not an obvious change, and a small vision deficiency makes hard to find why the webservice is not reachable on its usual port, which one would expect, if the package declares to provide httpd/nginx.

The `kaltura-nginx` package can be used for multiple purposes: it can serve as a standalone Nginx with the VOD module statically compiled but we also use it for Kaltura CE (see working modes section here https://github.com/kaltura/nginx-vod-module/#features and https://github.com/kaltura/platform-install-packages). Since, in CE context, depending on node role assignments, we may have Apache on the same machine, using 80 and 443, respectively, the default Nginx ports for the `kaltura-nginx` package are 88 and 8443. However, the user is prompted about these selections during the `configure` phase so, I do not regard this as an issue.

You can see current configuration selections with:
> # debconf-get-selections |grep kaltura-nginx

> and at least symlink the nginx binary to /usr/sbin/nginx
Third party packages should ideally not place files under "system" paths. We try to adhere to https://www.debian.org/doc/packaging-manuals/fhs/fhs-3.0.html#optAddonApplicationSoftwarePackages
Naturally, if you want to create a symlink, you're welcome to do so.

Cheers,

**Comment by gnanet** [comment]
Thakns for the extensive answer. The "what and why" is this way at least is described, so others may find this. As it is basically not an unsolvable task to apply own modifications, i rhink we can close this.
